src/basic/audit-util.c

```diff
@@ -68,6 +70,75 @@ int audit_loginuid_from_pid(pid_t pid, uid_t *uid) {
         return 0;
 }
 
+static int try_audit_request(int fd) {
+        struct nlmsgerr *nlerr = NULL;
+        struct nlmsghdr nlh;
+        struct iovec iov;
+        struct msghdr mh;
+        struct sockaddr_nl nladdr;
+        ssize_t n;
+
+        nlh = (struct nlmsghdr) {
+                .nlmsg_len = NLMSG_LENGTH(0),
+                .nlmsg_type = AUDIT_GET_FEATURE,
+                .nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK,
+        };
+        iov = (struct iovec) {
+                .iov_base = &nlh,
+                .iov_len = NLMSG_LENGTH(0),
+        };
+        mh = (struct msghdr) {
+                .msg_iov = &iov,
+                .msg_iovlen = 1,
+        };
+
+        if (sendmsg(fd, &mh, MSG_NOSIGNAL) < 0) {
+                log_debug_errno(errno, "Failed to send AUDIT_GET_FEATURE request, ignoring: %m");
+                return 0;
+        }
+
+        nlh = (struct nlmsghdr){};
+        iov = (struct iovec) {
+                .iov_base = &nlh,
+                .iov_len = NLMSG_LENGTH(sizeof(struct nlmsgerr)),
+        };
+        mh = (struct msghdr) {
+                .msg_name = &nladdr,
+                .msg_namelen = sizeof(nladdr),
+                .msg_iov = &iov,
+                .msg_iovlen = 1,
+        };
+
+        n = recvmsg_safe(fd, &mh, 0);
+        if (n < 0) {
+                log_debug_errno(errno, "Failed to recv AUDIT_GET_FEATURE request ack, ignoring: %m");
+                return 0;
+        }
+
+        if (!NLMSG_OK(&nlh, n)) {
+                log_debug("AUDIT_GET_FEATURE reply too large, ignoring.");
+                return 0;
+        }
+
+        if (nlh.nlmsg_type != NLMSG_ERROR) {
+                log_debug("Expected NLMSG_ERROR message but got %d, ignoring.", nlh.nlmsg_type);
+                return 0;
+        }
+
+        /* If we try and use the audit fd but get ECONNREFUSED, it is because
+         * we are not in the initial user namespace, and the kernel does not
+         * have support for audit outside of the initial user namespace. */
+        nlerr = NLMSG_DATA(&nlh);
+        if (nlerr->error == 0)
+                return 1;
+        else if (nlerr->error == -ECONNREFUSED)
```

Comment:
hmm, isn't the error returned here positive?

Reply:
the "else" is redundant

Reply:
I am getting a negative value (https://elixir.bootlin.com/linux/latest/source/kernel/audit.c#L1036), and looking at the relevant netlink code (https://elixir.bootlin.com/linux/latest/source/net/netlink/af_netlink.c#L2490) I don't see anything that tries to make this positive-valued.